tests/interfaces/test_autograd.py

```diff
@@ -475,3 +475,234 @@ def d_classical(a, b, method):
+        assert res[1].shape == weights2.shape
+
+        # check that the parameter shift was only performed for the
+        # differentiable input, not the data input
+        diff_variables = spy.call_args[1]["wrt"]
+        expected_diff_variables = set(data1.size + np.arange(weights1.size))
+        expected_diff_variables.update(data1.size + weights1.size + data2.size + np.arange(weights2.size))
+        assert diff_variables == expected_diff_variables
+
+    def test_gradient_non_differentiable_exception(self):
+        """Test that an exception is raised if non-differentiable data is
+        differentiated"""
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev, interface="autograd")
+        def circuit(data1):
+            qml.templates.AmplitudeEmbedding(data1, wires=[0, 1])
+            return qml.expval(qml.PauliZ(0))
+
+        grad_fn = qml.grad(circuit, argnum=[0])
+        data1 = qml.numpy.array([0, 1, 1, 0], requires_grad=False) / np.sqrt(2)
+
+        with pytest.raises(qml.numpy.NonDifferentiableError, match="is non-differentiable"):
+            grad_fn(data1)
+
+    def test_no_differentiable_parameters(self):
+        """If there are no differentiable parameters, the output of the gradient
+        function is an empty tuple, and a warning is emitted."""
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev, interface="autograd")
+        def circuit(data1):
+            qml.templates.AmplitudeEmbedding(data1, wires=[0, 1])
+            return qml.expval(qml.PauliZ(0))
+
+        grad_fn = qml.grad(circuit)
+        data1 = qml.numpy.array([0, 1, 1, 0], requires_grad=False) / np.sqrt(2)
+
+        with pytest.warns(UserWarning, match="Output seems independent of input"):
+            res = grad_fn(data1)
+
+        assert res == tuple()
+
+    def test_chained_qnodes(self):
+        """Test that the gradient of chained QNodes works without error"""
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev, interface="autograd")
+        def circuit1(weights):
+            qml.templates.StronglyEntanglingLayers(weights, wires=[0, 1])
+            return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))
+
+        @qml.qnode(dev, interface="autograd")
+        def circuit2(data, weights):
+            qml.templates.AngleEmbedding(data, wires=[0, 1])
+            qml.templates.StronglyEntanglingLayers(weights, wires=[0, 1])
+            return qml.expval(qml.PauliX(0))
+
+        def cost(weights):
+            w1, w2 = weights
+            c1 = circuit1(w1)
+            c2 = circuit2(c1, w2)
+            return qml.numpy.sum(c2) ** 2
+
+        w1 = qml.init.strong_ent_layers_normal(n_wires=2, n_layers=3)
+        w2 = qml.init.strong_ent_layers_normal(n_wires=2, n_layers=4)
+
+        weights = [w1, w2]
+
+        res = cost(weights)
+        grad_fn = qml.grad(cost)
+        res = grad_fn(weights)
+
+        assert len(res[0]) == 2
+
```

Comment:
anything more worth asserting here?

Reply:
Perhaps it's best to check that the gradients are correct, as we don't have any chained QNodes integration tests anywhere.